src/ray/raylet/worker_pool.h

```diff
@@ -39,6 +40,56 @@ namespace raylet {
 using WorkerCommandMap =
     std::unordered_map<Language, std::vector<std::string>, std::hash<int>>;
 
+/// \struct WorkerCacheKey
+///
+/// Struct used to cache workers, keyed by runtime_env. The dynamic_options and
+/// override_environment_variables will be unified into runtime_env in the future.
+struct WorkerCacheKey {
+  std::vector<std::string> dynamic_options;
+  std::unordered_map<std::string, std::string> override_environment_variables;
+  std::string serialized_runtime_env;
+  mutable std::size_t hash_ = 0;
+
+  bool operator==(const WorkerCacheKey &k) const {
+    return dynamic_options == k.dynamic_options &&
+           override_environment_variables == k.override_environment_variables &&
+           serialized_runtime_env == k.serialized_runtime_env;
+  }
+
+  bool EnvIsEmpty() const {
+    return dynamic_options.size() == 0 && override_environment_variables.size() == 0 &&
+           (serialized_runtime_env == "" || serialized_runtime_env == "{}");
+  }
+
+  std::size_t Hash() const {
+    // Cache the hash value.
+    if (!hash_) {
```

Comment:
This basically means the class itself should be immutable right (otherwise, caching could be broken)? Can we make all internal members const in this case, and instantiate them using the constructor initializer?